### PR TITLE
`Integrated code lifecycle`: Limit build logs size

### PR DESCRIPTION
**Pull request by BBesrour**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

There is an issue where an exercise can generate huge build logs, which when transferred over hazelcast, could crash the system with out of memory errors. This PR limits the build logs size to max 10.000 logs and max 1.024 chars per log.

### Description

Before sending the buildlogs over hazelcast, we truncate them to max 10.000 logs and 1.024 chars per log. Numbers are configurable via application properties (yml settings of the build agents)

**Please note that a log can consist of multiple lines. So it's normal if the lines per log file exceed 10000. But each log should not have more than 1024 chars. The number of logs can be counted by counting the timestamps (Ctrl+F `[Europe/Berlin]`)**

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Submit a few times as a student
2. As instructor, navigate to the build logs (/course-management/{courseId}/build-queue)
3. Make sure that you can view the build logs. Make sure that each build job has max 10000 logs (the number of logs is equal to the number of timestamps, you can figure out how logs there are by Ctrl+F `[Europe/Berlin]` and view count)

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Performance Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
	- Introduced configurable limits on the number of log entries and the length of each log entry for better log management.
	- Added a new `BuildLogDTO` class to improve the structure of log entries.

- **Bug Fixes**
	- Enhanced logging behavior during build job execution, ensuring logs are conditionally added based on new constraints.

- **Refactor**
	- Updated logging methods across multiple services to streamline log entry management and improve readability, transitioning from `BuildLogEntry` to `BuildLogDTO`.
	- Improved handling of build logs in various services to ensure consistency and efficiency in log management.

### CODE REVIEW

File: src/main/java/de/tum/cit/aet/artemis/programming/service/BuildLogEntryService.java [hunk 1]
```diff
@@ -323,8 +324,8 @@ public void saveBuildLogsToFile(List<BuildLogEntry> buildLogEntries, String buil
         Path logPath = exerciseLogsPath.resolve(buildJobId + ".log");
 
         StringBuilder logsStringBuilder = new StringBuilder();
-        for (BuildLogEntry buildLogEntry : buildLogEntries) {
-            logsStringBuilder.append(buildLogEntry.getTime()).append("\t").append(buildLogEntry.getLog());
+        for (var buildLogEntry : buildLogEntries) {
```
Comment (az108) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/BuildLogEntryService.java:327]: could we not use var here ?

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildLogsMap.java [hunk 2]
```diff
@@ -8,30 +8,77 @@
...
+    /**
+     * Appends a new build log entry to the build logs for the specified build job ID.
+     * Only the first maxCharsPerLine characters of the log message will be appended. Longer characters will be truncated to avoid memory issues.
+     * Only the first maxLogLinesPerBuildJob log entries will be stored. Newer logs will be ignored to avoid memory issues
+     *
+     * @param buildJobId the ID of the build job to append a log message to
+     * @param buildLog   the build log entry to append to the build log
+     */
+    public void appendBuildLogEntry(String buildJobId, BuildLogDTO buildLog) {
+        var buildLogs = buildLogsMap.computeIfAbsent(buildJobId, k -> new ArrayList<>());
```
Comment (az108) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildLogsMap.java:49]: could we not use var here to make the code clear

File: src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildJobContainerService.java [hunk 3]
```diff
@@ -414,7 +414,7 @@ private void executeDockerCommand(String containerId, String buildJobId, boolean
                 @Override
                 public void onNext(Frame item) {
                     String text = new String(item.getPayload());
-                    BuildLogEntry buildLogEntry = new BuildLogEntry(ZonedDateTime.now(), text);
+                    var buildLogEntry = new BuildLogDTO(ZonedDateTime.now(), text);
```
Comment (az108) [inline src/main/java/de/tum/cit/aet/artemis/buildagent/service/BuildJobContainerService.java:417]: same here with var

**Review by az108 — CHANGES_REQUESTED** [review]
Overall Code LGTM, only one small request for 3 spots

**Review by az108 — APPROVED** [review]
Thanks for the quick changes 👍

**Review by krusche — APPROVED** [review]

**Pull request merged**